vsphere/datadog_checks/vsphere/api.py

```diff
@@ -0,0 +1,175 @@
+# (C) Datadog, Inc. 2019-present
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+import functools
+import ssl
+
+from pyVim import connect
+from pyVmomi import vim, vmodl
+
+from datadog_checks.base import ensure_unicode, is_affirmative
+from datadog_checks.vsphere.constants import (
+    ALL_RESOURCES,
+    DEFAULT_BATCH_COLLECTOR_SIZE,
+    MAX_QUERY_METRICS_OPTION,
+    UNLIMITED_HIST_METRICS_PER_QUERY,
+)
+
+
+def smart_retry(f):
+    """A function decorated with this `@smart_retry` will trigger a new authentication if it fails. The function
+    will then be retried.
+    This is useful when the integration keeps a semi-healthy connection to the vSphere API"""
+
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        api_instance = args[0]
+        try:
+            return f(*args, **kwargs)
+        except Exception:
```

Comment:
I don't think we want anything specific here, the goal of that decorator really is to retry if anything goes wrong. If the error is not a flake it will be re-raised just after.